logs/replace.go

```diff
@@ -1,74 +1,41 @@
 package logs
 
 import (
+	"slices"
 	"sort"
 
 	"github.com/pganalyze/collector/state"
 )
 
-type logRange struct {
-	start int64
-	end   int64 // When working with this range, the character at the index of end is *excluded*
-}
-
-const replacementChar = 'X'
-
-// ReplaceSecrets - Replaces the secrets of the specified kind with the replacement character in the text
-func ReplaceSecrets(input []byte, logLines []state.LogLine, filterLogSecret []state.LogSecretKind) []byte {
-	var goodRanges []logRange
+const replacement = "[redacted]"
 
+func ReplaceSecrets(input []byte, logLines []state.LogLine, filterLogSecret []state.LogSecretKind) {
 	filterUnidentified := false
 	for _, k := range filterLogSecret {
 		if k == state.UnidentifiedLogSecret {
 			filterUnidentified = true
 		}
 	}
-
-	for _, logLine := range logLines {
-		goodRanges = append(goodRanges, logRange{start: logLine.ByteStart, end: logLine.ByteContentStart})
-		if logLine.ReviewedForSecrets {
+	for idx, logLine := range logLines {
+		if filterUnidentified && logLines[idx].Classification == 0 {
+			logLines[idx].Content = replacement + "\n"
+		} else {
+			content := input[logLines[idx].ByteContentStart:logLines[idx].ByteEnd]
 			sort.Slice(logLine.SecretMarkers, func(i, j int) bool {
 				return logLine.SecretMarkers[i].ByteStart < logLine.SecretMarkers[j].ByteEnd
 			})
-
-			// We're creating a good range when we find a filtered secret or the end (everything before is marked as good)
-			nextIdxToEvaluate := logLine.ByteContentStart
+			bytesChecked := 0
+			offset := 0
 			for _, m := range logLine.SecretMarkers {
-				filter := false
 				for _, k := range filterLogSecret {
-					if m.Kind == k {
-						filter = true
+					if m.Kind == k && m.ByteStart > bytesChecked {
+						content = slices.Replace(content, m.ByteStart-offset, m.ByteEnd-offset, []byte(replacement)...)
+						bytesChecked = m.ByteEnd
+						offset += max(1, len(replacement)-m.ByteEnd-m.ByteStart)
```

Comment:
I don't quite follow this. Should this be `len(replacement)-(m.ByteEnd-m.ByteStart)`? Why does the offset have a minimum of 1? Shouldn't a negative offset be possible if we're replacing very long content with a short marker?

Reply:
Looks like it actually needed to be `(m.ByteEnd - m.ByteStart) - len(replacement)`. I've updated the test to cover the case where long content is after a short marker.

Reply:
Ah yes, right. That looks good. Thanks for updating the test.